ansible_builder/main.py

```diff
@@ -71,13 +120,103 @@ def version(self):
+        if os.path.exists(meta_file):
+            with open(meta_file, 'r') as f:
+                self.raw = yaml.load(f)
+        else:
+            # A feature? Automatically infer requirements for collection
+            for entry, filename in [('python', 'requirements.txt'), ('system', 'bindep.txt')]:
+                candidate_file = os.path.join(collection_path, filename)
+                if os.path.exists(candidate_file):
+                    self.raw['dependencies'][entry] = filename
+
+    def target_dir(self):
+        namespace, name = self.namespace_name()
+        return os.path.join(
+            constants.base_collections_path, 'ansible_collections',
+            namespace, name
+        )
+
+    def namespace_name(self):
+        "Returns 2-tuple of namespace and name"
+        path_parts = [p for p in self.reference_path.split(os.path.sep) if p]
+        return tuple(path_parts[-2:])
+
     @property
-    def galaxy_requirements_file(self):
-        galaxy_file = self.raw.get('dependencies', {}).get('galaxy')
-        if galaxy_file is None or os.path.isabs(galaxy_file):
-            return galaxy_file
+    def python_requirements_relpath(self):
+        req_file = self.raw.get('dependencies', {}).get('python')
+        if req_file is None:
+            return None
+        elif os.path.isabs(req_file):
+            raise RuntimeError(
+                'Collections must specify relative paths for requirements files. '
+                'The file {0} specified by {1} violates this.'.format(
+                    req_file, self.reference_path
+                )
+            )
         else:
-            return os.path.join(os.path.dirname(self.filename), galaxy_file)
+            return req_file
+
+
+class UserDefinition(BaseDefinition):
+    def __init__(self, filename):
+        self.filename = filename
+        self.reference_path = os.path.dirname(filename)
+
+        try:
+            with open(filename, 'r') as f:
+                y = yaml.load(f)
+                self.raw = y if y else {}
+        except FileNotFoundError:
+            sys.exit("""
+            Could not detect '{0}' file in this directory.
+            Use -f to specify a different location.
+            """.format(constants.default_file))
+
+        self.manager = CollectionManager(self.galaxy_requirements_file)
+
+    def _get_dep_entry(self, entry):
+        deps = self.raw.get('dependencies', {})
+        req_file = deps.get(entry) if deps else None
+
+        if not req_file:
+            return None
+
+        if os.path.isabs(req_file):
+            return req_file
+
+        return os.path.join(self.reference_path, req_file)
+
+    @property
+    def python_requirements_file(self):
+        return self._get_dep_entry('python')
+
+    @property
+    def system_requirements_file(self):
+        return self._get_dep_entry('system')
+
+    @property
+    def galaxy_requirements_file(self):
+        return self._get_dep_entry('galaxy')
```

Comment:
These are still fairly redundant and verbose properties, if I do more cleanup before merging I would try to axe a lot of it.